osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour.Data;
+using osu.Game.Rulesets.Taiko.Objects;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour
+{
+    /// <summary>
+    /// Utility class to perform various encodings.
+    /// </summary>
+    public static class TaikoColourDifficultyPreprocessor
+    {
+        /// <summary>
+        /// Processes and encodes a list of <see cref="TaikoDifficultyHitObject"/>s into a list of <see cref="TaikoDifficultyHitObjectColour"/>s,
+        /// assigning the appropriate <see cref="TaikoDifficultyHitObjectColour"/>s to each <see cref="TaikoDifficultyHitObject"/>,
+        /// and pre-evaluating colour difficulty of each <see cref="TaikoDifficultyHitObject"/>.
+        /// </summary>
+        public static void ProcessAndAssign(List<DifficultyHitObject> hitObjects)
+        {
+            List<CoupledColourEncoding> encodings = encode(hitObjects);
+
+            // Assign indexing and encoding data to all relevant objects. Only the first note of each encoding type is
+            // assigned with the relevant encodings.
+            foreach (var coupledEncoding in encodings)
+            {
+                coupledEncoding.Payload[0].Payload[0].EncodedData[0].Colour.CoupledColourEncoding = coupledEncoding;
+
+                // The outermost loop is kept a ForEach loop since it doesn't need index information, and we want to
+                // keep i and j for ColourEncoding's and MonoEncoding's index respectively, to keep it in line with
+                // documentation.
+                for (int i = 0; i < coupledEncoding.Payload.Count; ++i)
+                {
+                    ColourEncoding colourEncoding = coupledEncoding.Payload[i];
+                    colourEncoding.Parent = coupledEncoding;
+                    colourEncoding.Index = i;
+                    colourEncoding.Payload[0].EncodedData[0].Colour.ColourEncoding = colourEncoding;
+
+                    for (int j = 0; j < colourEncoding.Payload.Count; ++j)
+                    {
+                        MonoEncoding monoEncoding = colourEncoding.Payload[j];
+                        monoEncoding.Parent = colourEncoding;
+                        monoEncoding.Index = j;
+                        monoEncoding.EncodedData[0].Colour.MonoEncoding = monoEncoding;
```

Comment:
Nit, but I really dunno why these backlinks are set up only after the "encoding" process is done. It seems like it would be *much* simpler to do this *while* the encoding is running (for each encoding pass, iterate down to hitobjects, and assign the current "encoding object" to the difficulty object).

Reply:
The original version using pre-evaluated difficulties for each encoding type actually didn't require these back links at all, but it was decided to keep evaluator stateless and not have pre-evaluation in an earlier review. These back links were assigned here as I didn't want to further complicate the encoding processes further. I have made a quick implementation with them moved into the respective encoding passes [here](https://github.com/vunyunt/osu/tree/colour-encoding-test), but not sure if it's an improvement over this. If you think that would be preferrable or have a better idea of implementing it I'll push the changes to this branch.

Reply:
Let's just go with what's in here for now as the difference isn't as big as I'd expected to see.